src/d3a/d3a_core/sim_results/file_export_endpoints.py

```diff
@@ -370,50 +371,83 @@ def _accumulate_energy(self, area, child, is_cep: bool):
+            self.total_self_consumption += trade.offer.energy * 1000
+
+    def _accumulate_self_consumption_buffer(self, trade):
+        if trade.seller_origin in self.producer_list and trade.buyer_origin in self.ess_list:
+            self.self_consumption_buffer += trade.offer.energy * 1000
+
+    def _dissipate_self_consumption_buffer(self, trade):
+        if trade.seller_origin in self.ess_list and \
+                (trade.buyer_origin not in self.consumer_list or
+                 trade.buyer_origin not in self.ess_list) and \
+                self.self_consumption_buffer > 0:
+            if (self.self_consumption_buffer - trade.offer.energy * 1000) > 0:
+                self.self_consumption_buffer -= trade.offer.energy * 1000
+            else:
+                self.self_consumption_buffer = 0
+
+    def _accumulate_energy_trace(self):
+        for c_area in self.consumer_area_list:
+            for market in c_area.past_markets:
+                for trade in market.trades:
+                    self._accumulate_self_consumption(trade)
+                    self._accumulate_self_consumption_buffer(trade)
+                    self._dissipate_self_consumption_buffer(trade)
+
+    def _area_self_sufficiency(self, area):
+        self.producer_list = list()
+        self.consumer_list = list()
+        self.consumer_area_list = list()
+        self.ess_list = list()
+        self.total_energy_demanded = 0
+        self.total_self_consumption = 0
+        self.self_consumption_buffer = 0
+
+        self._accumulate_devices(area)
+
+        self._accumulate_energy_trace()
+
+        # in case when the area doesn't have any load demand
+        if self.total_energy_demanded <= 0:
+            self.self_sufficiency[area.name] = None
+            return {"self_sufficiency": None}
 
-        for markets in area.past_markets:
-            for trade in markets.trades:
-                # Total Electricity traded by house device
-                if trade.buyer in house_load_device:
-                    total_energy_bought += trade.offer.energy
-                # Electricity produced by PV
-                if trade.offer.seller in house_pv_device:
-                    trade_by_pv += trade.offer.energy
-                # PV electricity self_consumption
-                if trade.offer.seller in house_pv_device and trade.buyer in house_load_device:
-                    traded_from_pv += trade.offer.energy
-
-        if trade_by_pv != 0:
-            self_consumption_within_house = traded_from_pv / trade_by_pv
-        else:
-            self_consumption_within_house = 0
-        if total_energy_bought != 0:
-            self_sufficiency = traded_from_pv / total_energy_bought
-        else:
-            self_sufficiency = 0
+        """
+        In case when the ess' SOC at start of simulation was more than the end of simulation
+        due to which it will end-up discharging more than its initial state.
```

Comment:
Sorry, but this comment doesnt fit to the if block after (IMO).
If self_consumption_buffer > total_self_consumption, the Storage was charging, right?

Reply:
Improved the formula, have a look at later commit [at the lines below](https://github.com/gridsingularity/d3a/pull/572/commits/74d1bb3e0b76bd23583e1355d94c9dda363d731b#diff-6a3a50e2566925be315bd7728a1e6629R373).